Casper 2.0

kubaplas opened the pull request.

kubaplas committed:
Casper 2.0:

* Improved errors.
* Unified rust-toolchain across packages.
* Removed ContractPackageHash.
* Moved handling Odra errors outside rpc client.
* Removed casper types through odra types from rpc client.
* Update chainspec to be compatible with newest version of Condor.
* Native events.

kpob commented on odra-casper/wasm-env/src/host_functions.rs:
```diff
@@ -574,42 +577,24 @@ fn create_constructor_group(contract_package_hash: ContractPackageHash) -> URef
-        } => Address::try_from(contract_package_hash)
-            .map_err(|e| ApiError::User(ExecutionError::from(e).code()))
-            .unwrap_or_revert()
+fn caller_to_address(caller: Caller) -> Address {
```
Maybe replace with `From` implemented for `Adderss` directly?

kubaplas replied:
Done.

kpob commented on odra-casper/wasm-env/src/lib.rs:
```diff
@@ -21,10 +21,10 @@ pub use wasm_contract_env::WasmContractEnv;
-pub fn panic(_info: &core::panic::PanicInfo) -> ! {
-    core::intrinsics::abort();
-}
+// /// Panic handler for the WASM target architecture.
```
What happed with the panic handler?

kubaplas replied:
It is implemented by one of the casper crates already.

kpob commented on odra-casper/rpc-client/src/casper_client/configuration.rs:
```diff
@@ -66,18 +74,22 @@ impl CasperClientConfiguration {
-        panic!("{}", err)
-    })
-}
+    fn get_env_variable(name: &str) -> String {
```
I'd rather put it into `utils` module, or leave it as it was. No need to call with `Self::`

kubaplas replied:
What if I told you that I just deleted whole utils module?

zie1ony replied:
`odra-casper/rpc-client/src/utils.rs` is a thing, so maybe reuse the same method from there.

kpob commented on odra-casper/rpc-client/src/casper_client/configuration.rs:
```diff
@@ -66,18 +74,22 @@ impl CasperClientConfiguration {
 
-fn get_optional_env_variable(name: &str) -> Option<String> {
-    std::env::var(name).ok()
+    fn get_optional_env_variable(name: &str) -> Option<String> {
```
same here

kubaplas replied:
Same as above

kpob commented on odra-casper/rpc-client/src/casper_client.rs:
I would be careful about panicking in this module.  Working with livenet, a dozen times I've encountered a hard-to-debug panic. For instance, if the queried dictionary didn't exist. I would use `anyhow` and push errors to the pub functions.

zie1ony commented on core/src/address.rs:
```diff
@@ -82,35 +90,52 @@ impl Address {
+        match self {
+            Address::Account(_) => self.to_entity_addr().to_formatted_string(),
+            Address::Contract(package_hash) => {
+                PackageHash::new(package_hash.value()).to_formatted_string()
```
```suggestion
package_hash.to_formatted_string()
```

kubaplas replied:
Done

zie1ony commented on core/src/address.rs:
```diff
@@ -210,6 +233,15 @@ impl ToString for Address {
+    fn from(value: Address) -> Self {
+        match value {
+            Address::Account(account) => AddressableEntityHash::new(account.value()),
+            Address::Contract(contract) => AddressableEntityHash::new(contract.value())
```
Address can't be build from AddressableEntityHash, because it represents single contract or account, not a package.

kubaplas replied:
This code was removed completely.

zie1ony commented on core/src/callstack.rs:
```diff
@@ -209,18 +210,14 @@ mod tests {
-            Address::Contract(
-                ContractPackageHash::from_formatted_str(CONTRACT_PACKAGE_HASH).unwrap()
-            ),
+            Address::Contract(PackageHash::from_formatted_str(PACKAGE_HASH).unwrap()),
```
use `Address::new`

zie1ony commented on core/src/callstack.rs:
```diff
@@ -209,18 +210,14 @@ mod tests {
-            Address::Contract(
-                ContractPackageHash::from_formatted_str(CONTRACT_PACKAGE_HASH).unwrap()
-            ),
+            Address::Contract(PackageHash::from_formatted_str(PACKAGE_HASH).unwrap()),
```
use `Address::new`

zie1ony commented on core/src/list.rs:
```diff
@@ -178,187 +178,184 @@ impl<T: ToBytes + FromBytes + CLTyped + Default> List<T> {
     }
 }
 
-#[cfg(all(feature = "mock-vm", test))]
```
Interesting artefact.

zie1ony commented on examples/ourcoin/bin/build_schema.rs:
```diff
@@ -1,6 +1,4 @@
 #![doc = "Binary for building schema definitions from odra contracts."]
```
This file should look like this:

```rust
#![doc = "Binary for building schema definitions from odra contracts."]
#[allow(unused_imports)]
use ourcoin;

#[cfg(not(target_arch = "wasm32"))]
extern "Rust" {
    fn module_schema() -> odra::contract_def::ContractBlueprint;
    fn casper_contract_schema() -> odra::schema::casper_contract_schema::ContractSchema;
}

#[cfg(not(target_arch = "wasm32"))]
fn main() {
    odra_build::schema(unsafe { crate::module_schema() }, unsafe {
        crate::casper_contract_schema()
    });
}
```

zie1ony commented on examples/ourcoin/wasm/OurToken.wasm:
Remove wasm file from the repo. Update `.gitignore` so it not happens later again.

zie1ony commented on justfile:
```diff
@@ -6,16 +6,16 @@ BINARYEN_CHECKSUM := "c55b74f3109cdae97490faf089b0286d3bba926bb6ea5ed00c8c784fc5
 
+# TODO: reenable this   cd odra-casper/proxy-caller && cargo clippy --target=wasm32-unknown-unknown -- -D warnings -A clippy::single-component-path-imports
 clippy:
     cargo clippy --all-targets -- -D warnings
```
This should be rewritten, as examples, modules and benchmarks are now in workspace, so one call should be enough.

zie1ony commented on odra-casper/rpc-client/src/casper_client/configuration.rs:
```diff
@@ -1,20 +1,25 @@
     pub cspr_cloud_auth_token: Option<String>
 }
 
+#[cfg(feature = "std")]
```
Is this crate used without `std`?

kubaplas replied:
It will be used by js client.

zie1ony commented on odra-casper/test-vm/src/casper_host.rs:
```diff
@@ -64,9 +60,19 @@ impl HostContext for CasperHost {
 
     fn get_event(&self, contract_address: &Address, index: u32) -> Result<Bytes, EventError> {
+        if !contract_address.is_contract() {
+            panic!("Events can be retrieved only for contracts, not for accounts.")
```
Maybe return `Err(EventError)` instead of panicking.

zie1ony commented on odra-casper/test-vm/src/casper_host.rs:
```diff
@@ -64,9 +60,19 @@ impl HostContext for CasperHost {
     }
 
+    fn get_events_count(&self, contract_address: &Address) -> u32 {
+        if !contract_address.is_contract() {
```
same here

zie1ony commented on odra-casper/wasm-env/src/lib.rs:
```diff
@@ -20,11 +20,3 @@ pub use wasm_contract_env::WasmContractEnv;
 #[cfg(all(target_arch = "wasm32", not(feature = "disable-allocator")))]
 #[allow(unused_imports)]
 use ink_allocator;
-
```
Why is this not needed anymore?
Have you tried to compile with `disable-allocator`?

zie1ony commented on odra-macros/src/ast/wasm_parts.rs:
```diff
@@ -527,7 +548,9 @@ mod test {
                                 odra::casper_types::EntryPointAccess::Public,
-                                odra::casper_types::EntryPointType::Contract,
+                                odra::casper_types::EntryPointType::Called,
+                                odra::casper_types::EntryPointPayment::Caller,
```
What is this `Caller`?

zie1ony commented on odra-vm/src/vm/odra_vm_state.rs:
```diff
@@ -74,6 +74,7 @@ impl OdraVmState {
 
     pub fn emit_event(&mut self, event_data: &Bytes) {
         let contract_address = self.callstack.current().address();
+        #[allow(clippy::manual_inspect)]
```
Why?

kubaplas replied:
Newest clippy wants me to rewrite the code using inspect, but I failed...

zie1ony commented on odra-vm/src/vm/utils.rs:
```diff
@@ -22,6 +20,6 @@ pub fn contract_address_from_u32(i: u32) -> Address {
-    let account_str = format!("{}{}{}", "contract-package-", a, padding);
-    Address::Contract(ContractPackageHash::from_formatted_str(account_str.as_str()).unwrap())
+    let account_str = format!("{}{}{}", "package-", a, padding);
+    Address::Contract(PackageHash::from_formatted_str(account_str.as_str()).unwrap())
```
Use `Address::new`

zie1ony commented on odra-vm/src/vm/utils.rs:
```diff
@@ -1,7 +1,5 @@
+use odra_core::casper_types::PackageHash;
+use odra_core::{casper_types::account::AccountHash, Address};
 
 pub fn account_address_from_str(str: &str) -> Address {
```
Could this method live in `address.rs`?

kubaplas replied:
This code is used only by odra_vm, mostly for tests, so there is no point of littering address.

zie1ony commented on odra-casper/livenet-env/src/livenet_host.rs:
```diff
@@ -209,3 +218,21 @@ impl HostContext for LivenetHost {
     }
 }
+
+fn find_wasm_file_path(wasm_file_name: &str) -> PathBuf {
```
This lived in `casper_client.rs`, have you moved it her, or copied?

kubaplas replied:
Copied, now it is removed

zie1ony commented on odra-casper/proxy-caller/src/lib.rs:
```diff
@@ -185,3 +185,11 @@ fn to_ptr<T: ToBytes>(t: T) -> (*const u8, usize, Vec<u8>) {
+#[cfg(target_arch = "wasm32")]
+#[panic_handler]
+#[no_mangle]
+pub fn panic(_info: &core::panic::PanicInfo) -> ! {
```
We have to discuss this as well.

kubaplas committed:
Updated chainspec to be compatible with casper-execution-engine 7.0.3
Updated github actions to use the latest version of artifact management
actions.

kubaplas committed:
Merge pull request #512 from odradev/bugfix/chainspec-breaking-change

Updated chainspec to be compatible with casper-execution-engine 7.0.3

kpob committed:
Allow HostRef to modify odra_cfg when deploying the contract (#514)

* Remove unused file
* Update schema description
* Add OdraConfig trait to set `odra_cfg` properties
* Add a livenet example with `OdraConfig`

kpob committed:
Fail calling `init` on OdraVM (#516)

* Add execution context to contract container
* Add test of a manual `init` call

kpob committed:
Add more traits and structs to odra::prelude (#517)

* Add more traits and structs to odra::prelude
* fix test in odra-macros
* update templates
---------

Co-authored-by: Maciej Zieliński <maciej@odra.dev>

kpob committed:
Move odra cli to the main repo (#515)

* Move odra-cli to the main repo

---------

Co-authored-by: Maciej Zieliński <maciej@odra.dev>

zie1ony committed:
Changelog and version bump for 1.4.0 (#521)

kubaplas committed:
Rebased and using newest Casper code, still wip

kubaplas committed:
Fixed address conversion.

kubaplas committed:
Some fixes to livenet, but still waiting for client/nctl/node stability.

kubaplas committed:
Use ces from Odra repo
